dart/dynamics/JacobianNode.h

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_DYNAMICS_JACOBIANNODE_H_
+#define DART_DYNAMICS_JACOBIANNODE_H_
+
+#include <memory>
+#include <unordered_set>
+
+#include "dart/dynamics/Frame.h"
+#include "dart/dynamics/Node.h"
+#include "dart/dynamics/SmartPointer.h"
+
+namespace dart {
+namespace dynamics {
+
+class Skeleton;
+class DegreeOfFreedom;
+class InverseKinematics;
+
+/// The JacobianNode class serves as a common interface for BodyNodes and
+/// EndEffectors to both be used as references for IK modules. This is a pure
+/// abstract class.
+class JacobianNode : public virtual Frame, public virtual Node
+{
+public:
+
+  /// This constructor only exists as a formality, because Entity and Frame
+  /// require it. Since this class is pure abstract and virutally inherits
+  /// Frame, the most derived class needs to call the constructor of Entity and
+  /// Frame anyway.
+  JacobianNode();
+
+  /// Virtual destructor
+  virtual ~JacobianNode();
+
+  /// Get a pointer to an IK module for this JacobianNode. If _createIfNull is
+  /// true, then the IK module will be generated if one does not already exist.
+  const std::shared_ptr<InverseKinematics>& getIK(bool _createIfNull = false);
```

Comment:
In general, I want to the user to be able to compactly and programmatically decide whether the IK should be automatically created. In other words, I want to allow:

```
void useThisIK(const InverseKinematicsPtr& ik);
/* ... */
  // Somewhere else in the code. autocreate is a bool whose value is determined at runtime
  useThisIK(bodyNode->getIK(autocreate));
```

Instead of

```
void useThisIK(const InverseKinematicsPtr& ik);
/* ... */
  // Somewhere else in the code. autocreate is a bool whose value is determined at runtime
  if(autocreate)
    useThisIK(bodyNode->getOrCreateIK());
  else
    useThisIK(bodyNode->getIK());
```

I realize this example is a bit convoluted, but I'm having trouble thinking of a more concrete example.